Search API - stricter fieldname validation.

markharwood opened the pull request:
New "strict" flag on search results defaults to true and forces throwing of errors when all indices fail to find a field mapping referenced in search requests.
This will mean existing user searches that work today may not work in future versions.
Implementation details:
* Each shard now records all failed fieldmapper lookups in a set of fieldnames in the QueryShardContext. QueryShardContext is cloned per request so this looks like a convenient place to stash this information
* Each shard now returns set of unmapped fieldnames in the QuerySearchResult
* Reducer node logic in SearchPhaseController now records if all shard results agree a fieldname is unmapped
* FetchSearchPhase will now fail if strict search mode is on and there are fieldnames unanimously agreed as unmapped
* New integration test for unmapped field validation: UnmappedFieldValidationIT
* Many integrations tests altered to setStrict(false) in order to preserve tests of unmapped aggs etc when no fieldname validation

Closes #26013

markharwood commented:
This is an initial cut to make sure I'm on the right track. TODOs include:

* fixing YAML tests
* adding strict options to specialist "validate query" and "explain" apis too.
* Currently serves HTTP 503 response codes rather than 400s
* Docs for strict flag

markharwood commented:
Fixing some of the existing YAML tests like [this one](https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/20_terms.yml#L629) will be hard.

It has a "skip" section because it relies on a feature introduced in 5.2 so it skips from 0 to 5.2. It also relies on unmapped fields which this 7.0-targeted PR breaks meaning the test only should run from 5.2 to 6.9.99. Unfortunately test version ranges are described as "skip-if" periods not "run-if" periods so we can't express this condition. I can think of several ways around this:

1) Change the YAML test runners to allow multiple skip expressions (unless they already support this) and duplicate the existing tests for >=7.0 tests with added strict=false flags
2) Change the YAML test runners to allow a new "run-if-version" expressions (unless they already support this) and duplicate the existing tests for >=7.0 tests with the now-needed strict=false flags
3) Change the YAML test runners to allow strict=false settings in the YAML (ignoring them when targeting elasticsearch versions < 7.0 that don't support this new search flag)
4) Change the existing tests to create additional empty indices with the necessary unmapped fields to prevent errors being thrown in 7.0

None of these feel particularly great

jpountz commented on core/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -73,6 +73,8 @@
 
     private Boolean requestCache;
 
+    private Boolean strict = Boolean.TRUE;
```
this should be a `boolean`, not `Boolean`

jpountz commented on core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -567,11 +584,13 @@ private InternalAggregations reduceAggs(List<InternalAggregations> aggregationsL
                           InternalAggregations aggregations, SearchProfileShardResults shardResults, ScoreDoc[] scoreDocs,
                           SortField[] sortFields, DocValueFormat[] sortValueFormats, int numReducePhases, boolean isSortedByField, int size,
-                          int from, boolean isEmptyResult) {
+                          int from, boolean isEmptyResult, Set<String> unanimouslyUnmappedFieldNames) {
```
I see many places that need to null check this variable. Can we require that it is not null all the time and make old indices just pass `Collections.emptySet()`?

jpountz commented on core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java:
```diff
@@ -85,7 +85,13 @@ private static Throwable deduplicateCause(Throwable cause, ShardSearchFailure[]
+            if (cause instanceof ElasticsearchException) {
+                return ((ElasticsearchException) getCause()).status();
+            }
+            // otherwise when no unsuccessful shards, it means no active shards, so just return SERVICE_UNAVAILABLE
```
I am a bit worried that we have to modify that logic. Is there a way that we could do without it?

s1monw replied:
I think we can do this but please use `ExceptionsHelper.status(cause)` it handles also when cause is null

markharwood replied:
To be clear, the suggestion being made is to replace the above if statement with:

        if (shardFailures.length == 0) {
            return ExceptionsHelper.status(unwrapCause());
        }

Note that would change the error response of any non-check-field errors from a 503 status to a 500.
This would be a change to existing behaviour.

jpountz commented on core/src/main/java/org/elasticsearch/search/internal/ShardSearchLocalRequest.java:
```diff
@@ -72,6 +72,7 @@
     private float indexBoost;
     private SearchSourceBuilder source;
     private Boolean requestCache;
+    private Boolean strict;
```
let's make it a `boolean`?

jpountz commented on core/src/main/java/org/elasticsearch/search/internal/ShardSearchLocalRequest.java:
```diff
@@ -93,11 +95,12 @@ public ShardSearchLocalRequest(ShardId shardId, String[] types, long nowInMillis
         this.nowInMillis = nowInMillis;
         this.aliasFilter = aliasFilter;
         this.shardId = shardId;
+        //TODO default strict field here?
```
yes please

markharwood replied:
...or add a `strict` parameter? (constructor is currently called from ValidateQuery and Explain actions).
Incidentally, where would be a good home to store a global constant for DEFAULT_STRICT_POLICY?

jpountz commented on core/src/main/java/org/elasticsearch/search/internal/ShardSearchLocalRequest.java:
```diff
@@ -210,6 +219,9 @@ protected void innerReadFrom(StreamInput in) throws IOException {
         }
+        if (in.getVersion().onOrAfter(Version.V_6_0_0_beta1)) {
+            strict = in.readOptionalBoolean();
+        }
```
serialization in SearchPhaseExecutionException.java uses 7.0 while this class uses 6.0, which version will this feature get in?

markharwood replied:
Cut/paste error. Chatted with @colings86 and we agreed 6.1 - will add a separate note on this PR with more details on this decision.

jpountz commented on core/src/main/java/org/elasticsearch/search/query/QuerySearchResult.java:
```diff
@@ -356,4 +372,16 @@ public long getTotalHits() {
+
+    /*
+     * A set of the unmapped fieldnames referenced in this search request or null
+     * if all fieldnames were valid.
```
could we instead return an empty set if all field names are valid?

markharwood commented:
Following a conversation with @colings86 we agreed a couple of changes to this PR:

1) Target 7.0 with strict=true default and also 6.1 but with strict=false default. Helps with BWC on YAML tests as strict=false is implicit in 5.x and 6.1 and can be left unstated in tests and in the 7.0 versions of YAML tests "strict=true" flag will be accepted by both 6.x and 7.0
2) "strict" flag name is perhaps too specific (maybe we want to add general request-validation for deprecated APIs/clauses etc later?) so change to more focused "strict_fieldnames" flag

colings86 commented:
> Target 7.0 with strict=true default and also 6.1 but with strict=false default. Helps with BWC on YAML tests as strict=false is implicit in 5.x and 6.1 and can be left unstated in tests and in the 7.0 versions of YAML tests "strict=true" flag will be accepted by both 6.x and 7.0

Another reason for doing this is to improve the upgrade path for users because before upgrading to 7.0 the user can check their queries with the `strict=true` setting to make sure all their queries are valid before upgrading so they don't get unexpected errors after the upgrade.

jpountz commented:
:+1: The plan makes sense to me.

markharwood commented on core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java:
```diff
@@ -85,7 +85,13 @@ private static Throwable deduplicateCause(Throwable cause, ShardSearchFailure[]
+            if (cause instanceof ElasticsearchException) {
+                return ((ElasticsearchException) getCause()).status();
+            }
+            // otherwise when no unsuccessful shards, it means no active shards, so just return SERVICE_UNAVAILABLE
```
I was getting 500 errors despite deliberately throwing a 400-status ParsingException from the FetchSearchPhase.innerRun() method.
The general exception handling logic in place looks to assume that all exceptions thrown without a single shard failure must be an internal 500 error which is clearly not the case here. I toyed with the idea of making my strict test fabricating a single-shard failure and passing that through somehow but thought it might be better to change the central error handling to use whatever status flag might be set on the failing exception.

markharwood commented:
Agreed with @colings86 that we should make this `strict_fieldnames` mode default to "false" initially because there are too many reference docs that will fail tests due to unmapped field names and fixing these in master would complicate a 6.1 backport. The best approach is therefore:
1) push this PR to master with strict_fieldnames=false default
2) backport 1)  to 6.1 branch
3) Open a new issue to change master's `strict_fieldnames` default to true
4) Create a PR for 3) that also tidies up all the reference docs that will now fail with unmapped fields.

It looks like current tests that are failing are down to serialisation issues when testing against stale 6.1 snapshots. 
If there's no objections I'll squash+rebase on latest master and force push here ready for a final review.

colings86 commented on core/src/main/java/org/elasticsearch/action/admin/indices/validate/query/TransportValidateQueryAction.java:
```diff
@@ -154,12 +154,18 @@ protected ShardValidateQueryResponse shardOperation(ShardValidateQueryRequest re
+            if (request.strictFieldNames()) {
+                Set<String> unmappedFields = searchContext.getQueryShardContext().getUnmappedFields();
+                if (unmappedFields.size() > 0) {
+                    throw new ParsingException(null, "The following fields were unmapped: " + unmappedFields);
```
maybe we should say `The following fields were unmapped on all shards:` here since there could be other fields that are only unmapped on a subset of shards?

markharwood replied:
Having discussed this we agreed to add index name into the error message here.

colings86 commented on core/src/main/java/org/elasticsearch/search/query/QuerySearchResult.java:
```diff
@@ -306,6 +311,13 @@ public void readFromWithId(long id, StreamInput in) throws IOException {
+            String[] unmappedFieldsArray = in.readOptionalStringArray();            
+            unmappedFields = unmappedFieldsArray == null ? null : new HashSet<>(Arrays.asList(unmappedFieldsArray));
+        } else {
+            unmappedFields = null;
```
could we not make this an empty set and avoid null checks?

markharwood replied:
True but at the cost of adding some inefficiency.
We'd be allocating empty arrays for serialization of these empty sets then, when deserializing we're allocating HashSet objects to load these empty arrays (unless we add special-case code for spotting size 0 arrays and using empty set which is effectively as ugly as the existing null checks?)

colings86 replied:
Ok fair enough, I'm happy leaving this as is then

jpountz replied:
Actually I'd still prefer to go with Colin's idea to use empty sets. We can still optimize later by making sure to use a Collections.emptySet (which is a singleton) if the size is 0.

markharwood commented on core/src/main/java/org/elasticsearch/action/admin/indices/validate/query/TransportValidateQueryAction.java:
```diff
@@ -154,12 +154,18 @@ protected ShardValidateQueryResponse shardOperation(ShardValidateQueryRequest re
+            if (request.strictFieldNames()) {
+                Set<String> unmappedFields = searchContext.getQueryShardContext().getUnmappedFields();
+                if (unmappedFields.size() > 0) {
+                    throw new ParsingException(null, "The following fields were unmapped: " + unmappedFields);
```
I don't think we know at this stage that all shards failed because it's not a reduce-level operation like the test we have in the search reduction phase?

jpountz commented on core/src/test/java/org/elasticsearch/search/aggregations/UnmappedFieldValidationIT.java:
```diff
@@ -0,0 +1,121 @@
+        } catch (SearchPhaseExecutionException e) {
+            assertTrue(e.getCause().getMessage().contains(
+                    "The following fields were unmapped in all indices searched: [all_unmapped_str]"));
+        }
```
FYI we usually prefer expectThrows which avoids some of the traps of the try/fail/catch/assert idiom

jpountz commented:
I'd also like it better if the name of the option was more specific and had `mapped` in it since it is really validating whether fields are mapped? For instance I think we use `allow_unmapped` elsewhere for something similar?

markharwood commented:
@jpountz Thanks for the review.
What would be the desired behaviour if field is only mapped on a can_match-excluded index with strict_fieldnames turned on? I assume users would expect that a valid field mapping on the excluded indices would still cancel out any unmapped reports on the included indices because we don't expect users to be aware of the low-level optimisations we might make in choices of index.

markharwood commented:
>I'd also like it better if the name of the option was more specific and had mapped in it since it is really validating whether fields are mapped?

Might it be better as an enum -  allow_unmapped options = all, one, none?
Default today is allow_umapped:all and in future, "one".

jpountz commented:
>  I assume users would expect that a valid field mapping on the excluded indices would still cancel out any unmapped reports on the included indices because we don't expect users to be aware of the low-level optimisations we might make in choices of index.

Agreed.

> Might it be better as an enum - allow_unmapped options = all, one, none?
> Default today is allow_umapped:all and in future, "one".

Good question... I think "one" is not good since we actually mean "all but one"?

markharwood commented:
>Good question... I think "one" is not good since we actually mean "all but one"?

Good point. The "unmapped" negative makes it tricky to reason about. If you reversed it and made it min_mapped (as in min_should_match) then that can be "zero", "one" or "all".

Naming aside, what do we think to the enum idea in general though?

colings86 commented:
Personally I don't see anything other than all or none being useful. You either want to make sure all your fields are mapped or you don't. I don't see what it gives you if you allow a single shard to have an unmapped field.

markharwood commented:
> I don't see what it gives you if you allow a single shard to have an unmapped field.

min_mapped options explained:

**0** - I don't care if my field names are misunderstood by all indices (our current default)
**1** - I need at least one index to know what I'm talking about (future default)
**all** - I don't have or expect heterogenous index shapes. They all better understand what I'm talking about in the choice fieldnames.

colings86 commented:
ok I see, I think I got tripped up by the negatives in the naming there 😄

markharwood commented:
So do you think a tri-state setting is likely to be too confusing here?

markharwood commented:
@jpountz - we do have an issue currently with canMatch optimisations.

From SearchService.canMatch:

    This method does a very quick rewrite of the query and returns true 
    if the query can potentially match any documents.

Looking at the rewrite logic that does not attempt any FieldMapper look ups currently so there's no opportunity to weave in any field validation logic here - it would just make the rewrite logic more complex/heavyweight.

I'm not sure there's an easy win to be had here - maybe strict mode should be relaxed/rejected when there's index skipping?

markharwood commented:
Proposal to deal with the canMatch issue and the question over if tri-state strictness is required:

we opt for a single strictness boolean that when true will throw an error if _any_ of the searched indexes reports an unmapped field. This strictness setting will only be useful when searching heterogenous indexes but that is likely to be a lot of deployments.

jpountz commented:
Fine with me. I was hopeful with your idea of failing only if none of the queried indices had the mapping, but it looks like it is not practical. So +1 to just making it a boolean.

markharwood commented:
>So +1 to just making it a boolean.

Sorry to flip-flop on this, @jpountz but I had some doubts on our decision because adoption may be close to zero if we can't make strict checking a default.

The options we have are:

Option | description | issues
------- | ----------- | ------
1  | The last proposal - errors if _any_ index has unmapped. |Too strict a policy to enable by default (we expect some people to have heterogenous indices) so unlikely to ever be used?
2 | The current code - errors if _no_ index has field mapped. | Strict policy could be enabled by default but logic doesn't account for the canMatch optimisation and can give false errors
3 | The current code (error if no index mapped) but fixing canMatch logic | More complex and slower rewrite logic to check field mappings or perhaps return full index mappings. CanMatch is intended to be "light touch" so possibly a concern

So it comes down to a choice between no adoption, false errors and more complex/slower code.

There's not a good clear solution here. Perhaps option 2 is preferable because its downside of false errors may be less likely to occur and when it does default strictness can be disabled. The "canMatch" check currently only optimises for date ranges and arguably the time-based indexing scenario. In these scenarios each index is typically the same mapping, just a different day's content and is therefore less likely to report unmapped field names.

markharwood commented:
@nik9000 introduced a fourth option which we have decided to implement:

we employ a best-efforts approach to validating fieldnames where if canMatch mode skips any index then we abort any fieldname validation (because it is too hard to reason about skipped indices).
Given that the validation is now best-efforts only the current setting name of `strict_fieldnames` is perhaps a little misleading. We should change this to the slightly less formal-sounding `check_fieldnames`. This will remain a boolean where `false` would run no validation and `true` would perform checking but only fail if
a) There was no skipping of indices due to canMatch and
b) none of the indices had a user-supplied fieldname mapped.

We would document this behaviour as best-efforts without diving into the weeds of when it would fail to fail.

markharwood commented:
There is greater potential for false negatives (failing to reporting unmappeds) than I originally thought.
Regardless of canMatch settings, query rewrite optimizations on actively searched shards can still  rewrite to matchNone and bypass any of our FieldMapper validation. In these cases the shard will return an empty set of unmapped fields and when ANDed with other shards will nullify any attempts to report bad fields.
Take for example a typical 5 shard setup with only 4 docs in the index. A query containing bad fieldnames may be optimized to a matchNone on one shard without checking any of the aggs/query clauses etc so will return an empty unmapped fieldnames set. When this empty set is ANDed with the other shards that reported the bad fieldnames the result is a false negative - no error.
The solution is for fieldname sets from shards in the same index to first be ORed together and then finally these per-index fieldname sets are ANDed to get a list of globally unmapped fieldnames.

markharwood commented:
Additions to deal with the fuzzy nature of these checks:
* Switched from `strict_fieldnames` to `check_fieldnames`.
* Added tolerance for canMatch optimisations - we don't try error anything if we have skipped shards
* Reduce false negatives potential by ORing unmapped fieldname sets from same index then ANDing cross-index sets

markharwood commented:
Woot! Green checks at last. I commented out some of the stuff that was causing BWC failures with the current 6.1 snapshot builds and will have to reinstate them in master once this check_fieldnames support is back-ported into 6.x.
So the plan is:
* commit this PR to master.
* backport to 6.x
* re-enable BWC check_fields tests in master 
* open an issue to make check_fieldnames=true in master which will break many reference docs tests and work on fixing them.

markharwood commented:
@jpountz @colings86 can I trouble you for another review? I think this is ready now

colings86 commented on core/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ShardValidateQueryRequest.java:
```diff
@@ -95,6 +103,13 @@ public void readFrom(StreamInput in) throws IOException {
+        // TODO - reinstate this line once checkFieldNames is backported to 6.1 
+        // otherwise 7.0 commit will cause many BWC failures
+//        if (in.getVersion().onOrAfter(Version.V_6_1_0)) {
+        if (in.getVersion().after(Version.V_6_1_0)) {
```
We will need to make sure this line reads `if (in.getVersion().onOrAfter(Version.V_6_1_0)) {` in the version that is backported to 6.1 too right?

colings86 replied:
I'm actually wondering if it would be better to commit with the `onOrAfter` line and just accept the errors for a build or two. The last good commit stuff should mean that only the intake build fails on this. You could also set up the backport for 6.x branch before pushing the change on master so you can push both at the same time and minimise the chances of builds failing.

colings86 commented on core/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ShardValidateQueryRequest.java:
```diff
@@ -109,5 +124,11 @@ public void writeTo(StreamOutput out) throws IOException {
+        // TODO - reinstate this line once checkFieldNames is backported to 6.1 
+        // otherwise 7.0 commit will cause many BWC failures
+//        if (out.getVersion().onOrAfter(Version.V_6_1_0)) {
+        if (out.getVersion().after(Version.V_6_1_0)) {
```
as above

markharwood commented:
>We will need to make sure this line reads if (in.getVersion().onOrAfter(Version.V_6_1_0)) { in the version that is backported to 6.1 too right?

Correct. I wanted to minimise the build noise it was generating. While there is a gap between master and 6.x that will need to be closed I figured it was easier to keep master in an unbroken state while I work on the backport to 6.x. Once 6.x is in there's a simple commit to master to clean up the serialization and YML logic which concerns which versions of 6.x. support check_fieldnames.

markharwood commented:
@jpountz @colings86  - this is my proposed plan for committing these changes which is up for discussion:

# 1)  Commit this PR to master
All tests are currently passing 

# 2) Backport to  6.x
Backport this PR but with these changes...
*  Change serialization support from "after 6.1" to "on or after 6.1"  in `QuerySearchResult`,  `ShardValidateQueryRequest`, `ValidateQueryRequest`, `SearchRequest`, `ShardSearchLocalRequest`
* Set BWC check to >=6.1 in yaml test `140_pre_filter_search_shards.yml`
* Remove references to check_fieldnames:false from YML tests (`20_typed_keys.yml`, `20_terms.yml`)

# 3) Make master compatible with 6.x 
BWC tests from master ->6.x will be failing at this point. A quick fix:
*  Reinstate commented out "check_fieldnames:false" params passed in `20_typed_keys.yml`, `140_pre_filter_search_shards.yml, 20_terms.yml`
*  Change serialization support from "after 6.1" to "on or after 6.1"  in `QuerySearchResult`,  `ShardValidateQueryRequest`, `ValidateQueryRequest`, `SearchRequest`, `ShardSearchLocalRequest`

# 4) Open an new issue/PR on Master - change checkFieldnames default to true
Longer term objective which may require multiple people to help fix docs.
* Change DEFAULT_CHECK_FIELDNAMES to true in SearchRequest.java
* Change documentation of default in `docs/reference/search/uri-request.asciidoc`
* Add to breaking changes docs
* Fix the many broken reference docs CONSOLE tests.....

This approach means we have trivial changes to apply in step 3 to get the master -> 6.x BWC tests passing again after step 2.
If we don't have a step 3 we would have breaking builds while we apply the non-trivial set of changes in step 2.

markharwood commented:
@s1monw The one bit of this Adrien specifically wanted you to review was https://github.com/elastic/elasticsearch/pull/26115#discussion_r132961691

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -468,6 +473,16 @@ private ReducedQueryPhase reducedQueryPhase(Collection<? extends SearchPhaseResu
             QuerySearchResult result = entry.queryResult();
+            
+            // OR the list of unmapped fieldnames from the same index.
+            Set<String> unmappedFieldNamesForIndex = perIndexUnmappedFieldsNames.get(result.getSearchShardTarget().getIndex());
```
maybe just use `perIndexUnmappedFieldsNames.computeIfAbsent(result.getSearchShardTarget().getIndex(), key -> new HashSet())` instead?

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -495,6 +510,20 @@ private ReducedQueryPhase reducedQueryPhase(Collection<? extends SearchPhaseResu
         }
+        
+        // AND the sets of unmapped fields across multiple indices.
+        Set<String> unanimouslyUnmappedFieldNames = null;
```
maybe it would be simpler to just initialize `unanimouslyUnmappedFieldNames` with an `new HashSet`?

markharwood replied:
We need to have a first-time-round special condition otherwise ANDing with `retainAll` on an initially empty set always produces an empty final set. I'm using ==null as the special first time condition. Alternative would be to init the set with first index but I'm not sure that's less code.

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -604,6 +636,15 @@ private InternalAggregations reduceAggs(List<InternalAggregations> aggregationsL
+         * Throws an error if all indices have reported the same query field names as unmapped.
+         */
+        public void checkFieldNames() {
+            if (unanimouslyUnmappedFieldNames.size() > 0) {
```
maybe use `isEmpty() == false`

markharwood commented:
Thanks for the review, @s1monw 

> I wonder if we should force a canMatch phase ...  that way we can fail it earlier before we execute the heavy search?

Would that not penalize the majority to optimize a minority?
Long-term plan is to have `check_field_names` on-by-default and to throw hard errors. The objective here being to highlight programming errors in client apps.  Bad searches should be the exception and arguably we shouldn't slow down the majority of good searches with an extra phase.

markharwood commented:
jenkins test this

jpountz commented:
> I wonder if we should force a canMatch phase if we check field names which only does a rewrite and should already be able to tell if all fields are mapped.

When we started discussing this change, I was hoping we would be able to design something that has a low impact. Given that this suggestion essentially means that we plan on having a search phase to check whether fields are mapped, I'm on the fence. (That would also require us to call QueryBuilder.toQuery in the `can_match` phase since this is when field names are checked.)

Another thing that was important to me when we discussed this feature was that we would no longer be lenient by default, and only a tiny minority of use-cases would need to turn leniency on again. This is what made me like the suggestion to only fail if some fields are unanymously unmapped. But this comes with downsides: potential inconsistency if the field is only mapped on shards that are excluded by the `can_match` phase, and this raises the error quite late in the process.

Maybe we need to go back to a simpler implementation that fails when a single shard has a field that is unmapped. It's unclear to me how commonly users request several indices where a field is only mapped on a subset of indices. I wouldn't like to end up in a situation where Kibana needs to re-enable this leniency by default in order to be more user-friendly for instance. Maybe I'm just being paranoid and searches where only a subset of the indices have the field unmapped are very rare?

markharwood commented:
> It's unclear to me how commonly users request several indices where a field is only mapped on a subset of indices

I think this is common e.g. searching across log files for `remote_host` OR `host` fields when looking for an IP address. Different systems will store ip address info in different fields

jpountz commented:
Yeah, thinking more about it, I'm also worried about dynamic mappings and time-based indices. Newly created indices might miss a number of fields for a short period of time if all documents don't have exactly the same set of fields.

@s1monw Any opinion?

markharwood commented:
I'm not sure the downsides of the current approach are all that bad given the lack of a clear alternative.

> potential inconsistency if the field is only mapped on shards that are excluded by the can_match phase

Failing to fail on some queries is better than our status quo of failing to fail on _any_ queries.

> this raises the error quite late in the process.

Bad queries I'd expect to be a small percentage of all requests ( what sort of system tolerates repeated requests that produce no matches?). We shouldn't need to optimize for them.

javanna commented:
What is the status of this? Shall it be updated and integrated or closed? cc @elastic/es-search-aggs

colings86 commented:
@markharwood should we keep this open? I think we are still at a point where we aren't sure if this is something we should add and if so how it should work? Should we close this until we have conclusions on those discussions as part of #26013?

markharwood commented:
If I recall correctly we decided the costs/benefits didn't stack up on this.
* The costs were added complexity in the code base, each search having to bring back lists of missing  field names in order to determine if a search was universally misunderstood across all indices.
* The benefits were nice in theory (query validation) but couldn't be checked for all searches because we couldn't validate when a `can_match` optimisation skipped execution of one of the indices.
